### PR TITLE
CDRIVER-2064 Add bson_iter_init_from_data

### PR DESCRIPTION
**Pull request by xdg**

### CODE REVIEW

File: src/bson/bson-iter.c [hunk 1]
```diff
@@ -70,6 +70,51 @@ bson_iter_init (bson_iter_t *iter,  /* OUT */
...
+                          const uint8_t *data, /* IN */
+                          size_t length)       /* IN */
+{
+   BSON_ASSERT (iter);
+   BSON_ASSERT (data);
+
+   if (BSON_UNLIKELY ((length < 5) || (length > INT_MAX))) {
+      memset (iter, 0, sizeof *iter);
+      return false;
+   }
```
Comment (bjori) [inline src/bson/bson-iter.c]: I think we should add the same overflow checks here as we do in other bson creation functions.
e.g. making sure the bson `data` reports the same length as `length` provided

File: doc/bson_iter_init_from_data.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
+:man_page: bson_iter_init_from_data_
```
Comment (bjori) [inline doc/bson_iter_init_from_data.rst]: I think there is a trailing `_` ?

File: doc/bson_iter_init_from_data.rst [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+     if (bson_iter_init_from_data (&iter, data, length) && bson_iter_find (&iter, "_id") &&
+         BSON_ITER_HOLDS_OID (&iter)) {
+        bson_iter_oid (&iter, &oid);
+        bson_oid_to_string (&oid, oidstr);
+        printf ("%s\n", oidstr);
+     } else {
+        printf ("Document is missing _id.\n");
+     }
+  }
+
```
Comment (bjori) [inline doc/bson_iter_init_from_data.rst]: We added a way to automatically track/label all functions that do/result in similar way.
I think a crosslink between this function and bson_iter_init is well deserved to improve discovery of this function.

See `doc/bson_init.rst` and `doc/bson_init_from_json.rst`.

Let me know if you'd like to update the pr with it, or we can do it after merging it

File: src/bson/bson-iter.c (hunk 1 above)
Comment (xdg) [inline src/bson/bson-iter.c]: Done.

File: doc/bson_iter_init_from_data.rst (hunk 3 above)
Comment (xdg) [inline doc/bson_iter_init_from_data.rst]: I'd appreciate if you could do it after merging.  I can cargo cult it, but you can probably do it faster than explkain how to fix whatever I do wrong.

File: src/bson/bson-iter.c (hunk 1 above)
Comment (xdg) [inline src/bson/bson-iter.c]: Good point.  Done.

File: doc/bson_iter_init_from_data.rst (hunk 3 above)
Comment (ajdavis) [inline doc/bson_iter_init_from_data.rst]: Done:

https://github.com/mongodb/libbson/commit/b643cc3b38fcf3158a732aaf7b152ba089efa423

screen shot 2017-02-28 at 5 02 54 pm

**Pull request merged**